omod/src/main/java/org/openmrs/module/owa/web/controller/OwaRestController.java

```diff
@@ -102,126 +106,146 @@
-					if (zip.size() == 0) {
-						message = messageSourceService.getMessage("owa.blank_zip");
-						file.delete();
-						log.warn("Zip file is empty");
-						session.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, message);
-						response.sendError(500, message);
-					} else {
-						ZipEntry entry = zip.getEntry("manifest.webapp");
-						if (entry == null) {
-							message = messageSourceService.getMessage("owa.manifest_not_found");
-							log.warn("Manifest file could not be found in app");
-							file.delete();
-							session.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, message);
-							response.sendError(500, message);
-						} else {
-							String contextPath = request.getScheme() + "://" + request.getServerName() + ":"
-									+ request.getServerPort() + request.getContextPath();
-							appManager.installApp(file, fileName, contextPath);
-							file.delete();
-							message = messageSourceService.getMessage("owa.app_installed");
-							response.setStatus(200);
-							session.setAttribute(WebConstants.OPENMRS_MSG_ATTR, message);
-						}
-					}
+				Boolean checkInstall = this.installApp(file, fileName, request, response, session);
+				if (checkInstall){
+					appManager.reloadApps();
+					appList = appManager.getApps();
+					return new ResponseEntity<List<App>>(appList, HttpStatus.OK);
 				}
-				catch (Exception e) {
-					message = messageSourceService.getMessage("owa.not_a_zip");
-					log.warn("App is not a zip archive");
-					file.delete();
-					session.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, message);
-					response.sendError(500, message);
-				}
-			} else {
-				message = messageSourceService.getMessage("owa.invalid_url");
-				log.warn("Invalid url");
-				session.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, message);
-				response.sendError(400, message);
+				return new ResponseEntity<List<App>>(HttpStatus.valueOf(500));
+			}
+			catch (FileNotFoundException e) {
+				// TODO Auto-generated catch block
+				e.printStackTrace();
+				log.warn(e.getMessage());
+				session.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, e.getMessage());
+				response.sendError(400, e.getMessage());
+				return new ResponseEntity<List<App>>(HttpStatus.BAD_REQUEST);
 			}
-			appManager.reloadApps();
-			appList = appManager.getApps();
 		}
-		return appList;
+		return new ResponseEntity<List<App>>(HttpStatus.FORBIDDEN);
 	}
 
 	@RequestMapping(value = "/rest/owa/allowModuleWebUpload", method = RequestMethod.GET)
 	@ResponseBody
-	public boolean allowWebAdmin(HttpServletRequest request, HttpServletResponse response) {
-		 return ModuleUtil.allowAdmin();
+	public Boolean allowWebAdmin(HttpServletRequest request, HttpServletResponse response) {
+		return ModuleUtil.allowAdmin();
+	}
+
+	private Boolean installApp(File file, String fileName, HttpServletRequest request,
+		HttpServletResponse response, HttpSession session) throws IOException{
+		String message = null;
+		try (ZipFile zip = new ZipFile(file)) {
```

Comment:
Did you actually change all these? Or are they just formatting changes?

Reply:
since this action was being used both by install and upload. I separated it and created a method of it's own. This led to the multiple changes.

Reply:
This makes it hard to tell what actually changed. Can you fix the bug without doing any refactoring or moving around things? You can always create a new ticket and hence pull request for refactoring.

Reply:
Okay